Ready for Swift 6 blog post (#734)
swiftlang/swift-org-website: _posts/2024-06-20-ready-for-swift-6.md

---
layout: post
published: true
date: 2024-06-20 12:00:00
title: Plotting a Path to a Package Ecosystem without Data Race Errors
author: [daveverwer, svenaschmidt]
---

> Swift 6 introduces compile-time data race safety checking for any code that [opts in to use the Swift 6 language mode](https://www.swift.org/migration/documentation/swift-6-concurrency-migration-guide/swift6mode). While individual modules can adopt this mode incrementally and independently of their dependencies, the combined code (client and dependencies) achieves runtime data race safety only when all parts adopt it. Therefore, the quick adoption of Swift 6 language mode across the ecosystem of open-source packages is crucial in advancing data race safety across the entire Swift ecosystem.

Reviewer: It would be great to link "compile-time data race safety checking" to https://www.swift.org/migration/documentation/swift-6-concurrency-migration-guide/dataracesafety so that there's an easy way to get more information about what this means.

Author: Link added. Thanks!

> Swift 6 introduces compile-time data race safety checking for any code that [opts in to use the Swift 6 language mode](https://www.swift.org/migration/documentation/swift-6-concurrency-migration-guide/swift6mode). While individual modules can adopt this mode incrementally and independently of their dependencies, the combined code (client and dependencies) achieves runtime data race safety only when all parts adopt it. Therefore, the quick adoption of Swift 6 language mode across the ecosystem of open-source packages is crucial in advancing data race safety across the entire Swift ecosystem.

Reviewer: Suggest some minor tweaks to emphasize that partial adoption is of value and avoid giving any sense that this is existential for the success of Swift.
```suggestion
Swift 6 introduces compile-time data race safety checking for any code that [opts in to use the Swift 6 language mode](https://www.swift.org/migration/documentation/swift-6-concurrency-migration-guide/swift6mode). While individual modules can adopt this mode incrementally and independently of their dependencies, the full benefit of runtime data race safety is only realized when all modules have opted in. Therefore, the quick adoption of Swift 6 language mode across the ecosystem of open-source packages will play a key role in advancing data race safety across the entire Swift ecosystem.
```

## Tracking Swift 6 Readiness and Progress

> The Swift Package Index’s new “[Ready for Swift 6](https://swiftpackageindex.com/ready-for-swift-6)” page tracks progress toward data race safety across the entire package ecosystem. While packages can opt-in to Swift 6 language mode at their convenience, this page shows the number of packages that would pass those checks if all strict concurrency checks were enabled for all packages.

Reviewer 2: opt-in -> opt in

Author: Thanks! Fixed

> The Swift Package Index has been running these checks using Swift 6 nightly toolchains since early May, and you can already see the [number of packages without data race errors increasing](https://swiftpackageindex.com/ready-for-swift-6#total-zero-errors). Then, for the past week since the first beta version of Xcode 16 was unveiled the Swift Package Index build machines have been running fresh builds with the advantage of platform SDKs that have more Sendable conformance, bringing the total percentage of packages with zero data race errors to over 43%! That's a really great start for being only one week into the Xcode 16 beta process.

Reviewer 2: unveiled the -> unveiled, the

Author: Thanks! Fixed

> The Swift Package Index has been running these checks using Swift 6 nightly toolchains since early May, and you can already see the [number of packages without data race errors increasing](https://swiftpackageindex.com/ready-for-swift-6#total-zero-errors). Then, for the past week since the first beta version of Xcode 16 was unveiled, the Swift Package Index build machines have been running fresh builds with the advantage of platform SDKs that have more Sendable conformance, bringing the total percentage of packages with zero data race errors to over 43%! That's a really great start for being only one week into the Xcode 16 beta process.

Reviewer: ```suggestion
The Swift Package Index has been running these checks using Swift 6 nightly toolchains since early May, and there's already a steady reduction in the [number of packages with data race errors](https://swiftpackageindex.com/ready-for-swift-6#total-zero-errors). Then, for the past week since the first beta version of Xcode 16 was unveiled, the Swift Package Index build machines have been running fresh builds with the advantage of platform SDKs that have more `Sendable` conformance, bringing the total percentage of packages with zero data race errors to over 43%! That's a really great start for being only one week into the Xcode 16 beta process.
```

> The Swift Package Index has been running these checks using Swift 6 nightly toolchains since early May, and you can already see the [number of packages without data race errors increasing](https://swiftpackageindex.com/ready-for-swift-6#total-zero-errors). Then, for the past week since the first beta version of Xcode 16 was unveiled, the Swift Package Index build machines have been running fresh builds with the advantage of platform SDKs that have more Sendable conformance, bringing the total percentage of packages with zero data race errors to over 43%! That's a really great start for being only one week into the Xcode 16 beta process.

Reviewer: We probably shouldn't bring Xcode into the equation here -- this article is about Swift 6. Xcode happens to deploy an updated Swift toolchain, but the IDE itself doesn't do anything Swift-related, does it?

Author: You're correct, but I wanted to mark the occasion of it being the first Swift 6 version that shipped to a wider audience than those developers who would grab a nightly version. We could mention the toolchain date (although I'm not 100% sure how to find that for the version that ships in Xcode).

Author: We could just make this a little more vague and say "Then, for the past week, the Swift Package Index build machines have been running fresh builds with the advantage of platform SDKs that have more Sendable conformance"

Removing "since the first beta version of Xcode 16 was unveiled" altogether.

Reviewer: That could work, or you could just be more explicit and say, "with the first developer betas of Xcode 16, Swift 6 is now available to a much wider audience. And..."

Author: I just reworked this paragraph a little to remove both mentions of Xcode and refer to it as "The Swift 6 beta process"

![Screenshot 2024-06-26 at 20 44 33@2x](https://github.com/swiftlang/swift-org-website/assets/5180/7956dd67-b73d-4eb1-9f26-8abea7d37434)

> The Swift Package Index has been running these checks using Swift 6 nightly toolchains since early May, and you can already see the [number of packages without data race errors increasing](https://swiftpackageindex.com/ready-for-swift-6#total-zero-errors). Then, for the past week since the first beta version of Xcode 16 was unveiled, the Swift Package Index build machines have been running fresh builds with the advantage of platform SDKs that have more Sendable conformance, bringing the total percentage of packages with zero data race errors to over 43%! That's a really great start for being only one week into the Xcode 16 beta process.

Reviewer: I'm finding it hard to connect the data point here (43% zero data race errors) to the chart above. Is there a version of the chart that could show this percentage trend explicitly?

I'm also a bit confused by what we're communicating here? Are the 57% with zero data race errors packages that have not adopted Swift 6? Or packages that have adopted Swift 6 but have more work to do in later betas?

Author: I believe the percentage is correct, as shown by the tooltips below, but you're correct that we don't explicitly state the percentage in the data point tooltip.

![Screenshot 2024-06-26 at 19 55 39@2x](https://github.com/swiftlang/swift-org-website/assets/5180/62e35303-dda9-41e8-996a-59a04d846a56)

If I were able to add that percentage to the tooltip, would that make this clearer?

Author: > I'm also a bit confused by what we're communicating here? Are the 57% with zero data race errors packages that have not adopted Swift 6? Or packages that have adopted Swift 6 but have more work to do in later betas?

I agree this is a little confusing. This chart plots the status of packages as if they had been "forced" to adopt Swift 6 language mode and "complete" concurrency checking. It does not show how many packages have voluntarily opted in. We do explain this in FAQ question 2 below the chart, but we could potentially be clearer about that. I'll work on a revised version of the FAQ question.

Reviewer: > If I were able to add that percentage to the tooltip, would that make this clearer?

I wonder if the chart just needs inverting. Perhaps it's more natural to show "% of packages with data race errors" and show that the number is trending to 0%?

Reviewer: > This chart plots the status of packages as if they had been "forced" to adopt Swift 6 language mode and "complete" concurrency checking.

I'd just say that -- I think that's helpful context!

Author: I have updated the FAQ on the SPI Ready for Swift 6 page to include wording similar to this.

![Screenshot 2024-06-28 at 08 45 55@2x](https://github.com/swiftlang/swift-org-website/assets/5180/649cdd0f-0749-431f-b45e-cb6eadb4e39e)

![Chart showing the number of packages with no data race errors increasing over time from May this year](/assets/images/ready-for-swift-6-blog/packages-with-no-data-race-errors.png)

> Over time, successful adoption of Swift 6 language mode in the package ecosystem will come from two directions. First, as more packages opt into these checks, package authors will fix potential data races highlighted by the compiler, increasing compatibility and reducing errors. Second, compiler iterations will refine data race checking and consolidate error diagnostics.

Reviewer: What is the nature of the "refinement" in improved accuracy in the last sentence? Improving accuracy? Reducing false positives? Or something else?

Author: As I understand it, there are still some proposals in active review (like [SE-0430](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0430-transferring-parameters-and-results.md)) that will affect and potentially reduce the number of warnings/errors. I don’t know the exact status of all the other evolution proposals that are accepted, but it’s possible some are not yet in the compiler that shipped with beta1 or beta2. I don’t know, so I was intentionally a little vague here, but I know more is being done by Apple.

Reviewer: OK, vague is fine then. Thanks!

## Data Race Safety Indicators

> When evaluating a package, it is helpful to know if that package has any reported data race safety issues. To assist with this, the Swift Package Index now displays a “Safe from data races” label alongside other package metadata when packages compile with zero errors using `complete` strict concurrency checks.

Reviewer: Recommend link to https://www.swift.org/migration/documentation/swift-6-concurrency-migration-guide/completechecking to show developers how they can turn on this mode.

Author: Added a link here:

![Screenshot 2024-06-26 at 20 18 00@2x](https://github.com/swiftlang/swift-org-website/assets/5180/c063dbe5-46a3-4f72-86d7-0e4b298fcac4)

Author: > Great -- or just link the "complete strict concurrency checks" text to the URL, which will help with SEO!

Changed to this suggestion instead.

![Package metadata shown on the Swift Package Index showing that the package has no data race errors](/assets/images/ready-for-swift-6-blog/package-showing-safe-from-data-races.png)

### Compatibility vs Data Race Safety

> During the transition to Swift 6, you will likely see packages that show a green tick against Swift 6 in Swift Package Index’s platform and Swift version compatibility matrix while also showing that the package has data race safety errors:

Reviewer: Finish the thought here: what does that mean? Can the package be used from a project that has strict concurrency mode enabled? Is this a cause for concern that suggests the package is unstable and may crash? Or is the package still fine for adoption? 

Out of curiosity -- should we not have a different symbol for these packages? A yellow symbol? Or a lighter green checkmark?

Author: > Finish the thought here: what does that mean? Can the package be used from a project that has strict concurrency mode enabled? Is this a cause for concern that suggests the package is unstable and may crash? Or is the package still fine for adoption?

This made me realise I haven’t actually said what problems data races cause, so I added a sentence at the top of the article to cover that. I also covered whether the package can be used at the end of this section.

> Out of curiosity -- should we not have a different symbol for these packages? A yellow symbol? Or a lighter green checkmark?

I don’t think we should do this. The compatibility matrix has a clear job, and that is to report compatibility at a glance. Whether someone wants to use a package is based on lots of metadata in the above section, with data race safety being part of that, and compatibility being another part. We shouldn’t try to represent any other metadata with the compatibility information.

Reviewer: OK, sounds good! 

Unrelated to this thread, we saw good success with using scores to incentivize best practices for Flutter; see for example, https://pub.dev/packages/path_provider/score. Might be something to consider in the future for Swift Package Index?

Author: We have some work in this area but are deliberately not pushing it prominently yet as we did get a little push back when we implemented some of the metrics, and there are not enough yet to make it comprehensive. You can see the work we did on any "maintainer's" page, like this:

https://swiftpackageindex.com/SwiftPackageIndex/SemanticVersion/information-for-package-maintainers#package-score

I'd love to pick your brains sometime about how this worked for Flutter and how you rolled it out with the community.

![Screenshot 2024-06-28 at 19 24 03@2x](https://github.com/swiftlang/swift-org-website/assets/5180/304256cd-f504-4074-ac8f-bbcbe0d6e0fc)

Reviewer: Cool -- yeah, let's schedule some time to talk about this. I'll add some comments to https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/discussions/2591

> During the transition to Swift 6, you will likely see packages that show a green tick against Swift 6 in Swift Package Index’s platform and Swift version compatibility matrix while also showing that the package has data race safety errors:

Reviewer: American English is "checkmark", not "tick". I believe we standardize on US English for the blog, and the latter may not be so well understood by a broad audience.

Author: Good catch! 🇬🇧 Fixed

![A Swift Package Index compatibility matrix showing a green tick against Swift 6 compatibility](/assets/images/ready-for-swift-6-blog/package-compatibility-with-swift-6.png)

When compiling packages to check compatibility with Swift 6, the Swift Package Index uses whichever language mode the package author specifies. A package can be compatible with Swift 6 before it adopts the Swift 6 language mode and strict concurrency checks.

## Call to Action: Adopt the Swift 6 Language Mode

> Data race safety is a crucial feature of the Swift 6 compiler, eliminating many subtle or significant issues. For this feature to be effective, widespread adoption is essential. We urge package authors to [adopt Swift 6 readiness checks by following the Swift 6 migration guide](https://www.swift.org/migration/) to maximize data race safety across the ecosystem and follow along with the community’s progress on the [Ready for Swift 6 page on the Swift Package Index](https://swiftpackageindex.com/ready-for-swift-6).

Reviewer: Updating this language to draw from our carefully-honed State of the Union messaging!

```suggestion
Compile-time data race safety is a major advance for the Swift language, eliminating an entire class of potential concurrency bugs and elevating your code's safety and maintainability. Every module that migrates to Swift 6 contributes to the community-wide transition to bring data race safety to the Swift software ecosystem. You can help by [updating your own packages](https://www.swift.org/migration/), and you can follow the broader ecosystem's progress on the [Ready for Swift 6 page on the Swift Package Index](https://swiftpackageindex.com/ready-for-swift-6).
```

---
layout: post
published: true
date: 2024-06-20 12:00:00
title: Plotting a Path to a Package Ecosystem without Data Race Errors
author: [daveverwer, svenaschmidt]
---

Swift 6 introduces compile-time data race safety checking for any code that [opts in to use the Swift 6 language mode](https://www.swift.org/migration/documentation/swift-6-concurrency-migration-guide/swift6mode). While individual modules can adopt this mode incrementally and independently of their dependencies, the combined code (client and dependencies) achieves runtime data race safety only when all parts adopt it. Therefore, the quick adoption of Swift 6 language mode across the ecosystem of open-source packages is crucial in advancing data race safety across the entire Swift ecosystem.

## Tracking Swift 6 Readiness and Progress

> The Swift Package Index’s new “[Ready for Swift 6](https://swiftpackageindex.com/ready-for-swift-6)”  page tracks progress toward data race safety across the entire package ecosystem. While packages can opt-in to Swift 6 language mode at their convenience, this page shows the number of packages that would pass those checks if all strict concurrency checks were enabled for all packages.

Reviewer 3: ```suggestion
The Swift Package Index’s new “[Ready for Swift 6](https://swiftpackageindex.com/ready-for-swift-6)” page tracks progress toward data race safety across the entire package ecosystem. While packages can opt-in to Swift 6 language mode at their convenience, this page shows the number of packages that would pass those checks if all strict concurrency checks were enabled for all packages.
```

> The Swift Package Index’s new “[Ready for Swift 6](https://swiftpackageindex.com/ready-for-swift-6)” page tracks progress toward data race safety across the entire package ecosystem. While packages can opt in to Swift 6 language mode at their convenience, this page shows the number of packages that would pass those checks if all strict concurrency checks were enabled for all packages.

Reviewer: Do we need quotes around "Ready for Swift 6"? Consider removing.

Author: Removed. Thanks!

> The Swift Package Index has been running these checks using Swift 6 nightly toolchains since early May, and you can already see the [number of packages without data race errors increasing](https://swiftpackageindex.com/ready-for-swift-6#total-zero-errors). Then, for the past week since the first beta version of Xcode 16 was unveiled the Swift Package Index build machines have been runing fresh builds with the advantage of platform SDKs that have more Sendable conformance, bringing the total percentage of packages with zero data race errors to over 43%! That's a really great start for being only one week into the Xcode 16 beta process.

Reviewer 3: ```suggestion
The Swift Package Index has been running these checks using Swift 6 nightly toolchains since early May, and you can already see the [number of packages without data race errors increasing](https://swiftpackageindex.com/ready-for-swift-6#total-zero-errors). Then, for the past week since the first beta version of Xcode 16 was unveiled the Swift Package Index build machines have been running fresh builds with the advantage of platform SDKs that have more Sendable conformance, bringing the total percentage of packages with zero data race errors to over 43%! That's a really great start for being only one week into the Xcode 16 beta process.
```

> The Swift Package Index has been running these checks using Swift 6 nightly toolchains since early May, and there has already been a steady reduction in the [number of packages with data race errors](https://swiftpackageindex.com/ready-for-swift-6#total-zero-errors). For the past week, the Swift Package Index build machines have been running fresh builds with the advantage of platform SDKs with more `Sendable` conformance, bringing the total percentage of packages with zero data race errors to over 43%!That's a really great start for being only one week into the Swift 6 beta process.

Reviewer: Need a space here:
> over 43%!That's

Author: Fixed. Thanks!

> While packages can opt-in to Swift 6 language mode at their convenience, the chart shows the number of packages that would pass those checks if all strict concurrency checks were explicitly enabled for a package.

Reviewer 3: This line is very similar to the one on [line 13](https://github.com/swiftlang/swift-org-website/pull/734/files#diff-1f74d4e276a408a0d8dbf0732e7ebd6d6f4da648a46ba728b0142730a8fab036R13). Maybe we can rewrite/remove one of them?

>While packages can opt-in to Swift 6 language mode at their convenience, this page shows the number of packages that would pass those checks if all strict concurrency checks were enabled for all packages.

>While packages can opt-in to Swift 6 language mode at their convenience, the chart shows the number of packages that would pass those checks if all strict concurrency checks were explicitly enabled for a package.

Author: Well spotted. I think that second sentence adds nothing, I'll remove that one.

Over time, successful adoption of Swift 6 language mode in the package ecosystem will come from two directions. First, as more packages opt into these checks, package authors will fix potential data races highlighted by the compiler, increasing compatibility and reducing errors. Second, compiler iterations will refine data race checking and consolidate error diagnostics.

## Data Race Safety Indicators

> When evaluating a package, it is be helpful to know if that package has any reported data race safety issues. To assist with this, the Swift Package Index now displays a “Safe from data races” label alongside other package metadata when packages compile with zero errors using `complete` strict concurrency checks.

Reviewer 3: ```suggestion
When evaluating a package, it is helpful to know if that package has any reported data race safety issues. To assist with this, the Swift Package Index now displays a “Safe from data races” label alongside other package metadata when packages compile with zero errors using `complete` strict concurrency checks.
```
